shared/api/image_decoder_darwin.hpp

```diff
@@ -0,0 +1,121 @@
+#include <ImageIO/ImageIO.h>
+
+#include "op_def_struct.h"
+#include "ext_status.h"
+
+
+struct DecodeImage {
+  template <typename DictT>
+  OrtxStatus Init(const DictT& attrs) {
+    CFStringRef optionKeys[2];
+    CFTypeRef optionValues[2];
+    optionKeys[0] = kCGImageSourceShouldCache;
+    optionValues[0] = (CFTypeRef)kCFBooleanFalse;
+    // Only Integer image data is currently supported
+    optionKeys[1] = kCGImageSourceShouldAllowFloat;
+    optionValues[1] = (CFTypeRef)kCFBooleanFalse;
+
+    _imageSourceOptions = CFDictionaryCreate(NULL, (const void**)optionKeys, (const void**)optionValues, 2,
+                                             &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+
+    return {};
+  }
+
+  OrtxStatus Compute(const ortc::Tensor<uint8_t>& input, ortc::Tensor<uint8_t>& output) {
+    const auto& dimensions = input.Shape();
+    if (dimensions.size() != 1ULL) {
+      return {kOrtxErrorInvalidArgument, "[ImageDecoder]: Only raw image formats are supported."};
+    }
+
+    // Get data & the length
+    const uint8_t* encoded_image_data = input.Data();
+    const int64_t encoded_image_data_len = input.NumberOfElement();
+
+    // check whether it's too small for a image
+    if (encoded_image_data_len < 8) {
+      return {kOrtxErrorInvalidArgument, "[ImageDecoder]: Invalid image data."};
+    }
+
+    OrtxStatus status{};
+
+    CFDataRef imageData = NULL;
+    CGImageRef image = NULL;
+    CGImageSourceRef imageSource;
+
+    imageData = CFDataCreate(NULL, encoded_image_data, encoded_image_data_len);
+    if (imageData == nullptr) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: Failed to create CFData."};
+    }
+    imageSource = CGImageSourceCreateWithData(imageData, _imageSourceOptions);
+    CFRelease(imageData);
+
+    if (imageSource == nullptr) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: Failed to create CGImageSource."};
+    }
+
+    image = CGImageSourceCreateImageAtIndex(imageSource, 0, NULL);
+    CFRelease(imageSource);
+    if (image == nullptr) {
+      return {kOrtxErrorInternal, "[ImageDecoder]: Failed to create CGImage."};
+    }
+
+    const int64_t width = static_cast<int64_t>(CGImageGetWidth(image));
+    const int64_t height = static_cast<int64_t>(CGImageGetHeight(image));
+    const int64_t channels = 3;
+
+    std::vector<int64_t> output_dimensions{height, width, channels};
+    uint8_t* decoded_image_data = output.Allocate(output_dimensions);
+
+    if (decoded_image_data == nullptr) {
+      return {kOrtxErrorInvalidArgument, "[ImageDecoder]: Failed to allocate memory for decoded image data."};
+    }
+
+    // CoreGraphics don't support 24BPP. We get 32BPP with alpha first and then extract the 24BPP data we need.
+    const size_t _32bpp_channel = 4;
+    const size_t _32bpp_bytesPerRow = width * _32bpp_channel;
+    const size_t _32bpp_bitmapByteCount = width * height * _32bpp_channel;
+    void* _32bpp_bitmapData = malloc(_32bpp_bitmapByteCount);
```

Comment:
why malloc, instead of new? or can we use std::unique_ptr<..>?

Reply:
I just need raw memory area and `free` it after copying it to `output`.

Smart pointer would save us the `free` part. Do you mean like this `std::make_unique<int>(_32bpp_bitmapByteCount)`?

If there benefit from using `new int[_32bpp_bitmapByteCount]` and `delete` instead of malloc/free?

Reply:
Fixed with `std::make_unique`